### PR TITLE
pipeline: various improvements on sync, flush and fetch steps 

### PR DESCRIPTION
**Pull request by dlax**
(see commit messages for details; part of this would make sense with #271 as well)

### CODE REVIEW

**Comment by dvarrazzo** [comment]
I don't see tests here. Is this something addressing the problems found in #267?

Juggling a few things right now, I'll be back on this later

File: psycopg/psycopg/_pipeline.py [hunk 1]
```diff
@@ -196,7 +197,7 @@ def __exit__(
         exc_tb: Optional[TracebackType],
     ) -> None:
         try:
-            self.sync()
+            self.sync(True)
```
Comment (dvarrazzo) [inline psycopg/psycopg/_pipeline.py:200]: I would make this a kwarg-only argument, as `_sync_gen()` is,  and force user to use `fetch=True`.

**Comment by dlax** [comment]
> I don't see tests here. Is this something addressing the problems found in #267?

No, it's not related.

Rather it's a matter of avoiding surprises from user's perspective: "eh, I just wanted to sync, what's all this server->client network traffic?" (But the definition of "surprise" is subjective of course.)

> I don't see tests here.

I'd try, but I'm afraid it's going to be tricky to write reliable tests for this very change because, when we call `Pipeline.sync()`, we then call the `pipeline_communicate()` generator which might in turn fetch some results *if* the server has decided to flush its output buffer.

(In fact, this change reverts back to the original implementation I submitted where calling `sync()` did not involved explicit results fetching.)

File: psycopg/psycopg/_pipeline.py (hunk 1 above)
Comment (dlax) [reply psycopg/psycopg/_pipeline.py:200]: I'm not sure there is a use case for an explicit fetch so I'd rather rework the commit to not have the parameter in `sync()`. Do you see a use case on your side?

**Comment by dvarrazzo** [comment]
No worries about the tests. I was just wondering if this refactoring was something addressing the test failed that I introduced to show the rollback problem in #267

File: psycopg/psycopg/_pipeline.py (hunk 1 above)
Comment (dvarrazzo) [reply psycopg/psycopg/_pipeline.py:200]: No use case in mind. Just looking at that naked `True` in the `sync()`, which doesn't give any explanation, whereas the internal `_sync_gen()` is self explanatory with a kwargs-only argument :slightly_smiling_face:

**Comment by dlax** [comment]
Second commit is another (micro?) optimization on `executemany(..., returning=True)`.

**Comment by dvarrazzo** [comment]
Rebased on master, will merge as tests pass.

**Pull request merged**